### ISSUE TITLE
Projects with dependencies in CI

### ISSUE THREAD

**Issue by linusha** [LivelyKernel/lively.next#985]
This is something that I discovered while working on, https://github.com/LivelyKernel/froscon-2023, specifically the demo deployment:

We spent a good amount of time thinking about ensuring correct resolving and retrieving of dependencies locally, but this logic is implemented entirely inside of the client.
This means, that trying to bundle some project with a dependency in CI will fail, as there is nothing ensuring the existence of the dependency there. This uncovers quite a large pile of stuff that we entirely missed until now: **They all resolve around the questions of how to retrieve the necessary dependencies, while in CI.**

# Running Tests of Projects with dependencies in CI

I think this is the easiest case. This is, because to run the tests, we need to load a lively world anyways, meaning the dependencies get downloaded just like locally. _This leaves open the question on how an error retrieving the dependencies manifests itself in the test pipeline, but I'd think this is secondary, at least for now._

**In case we go ahead and implement private projects, this becomes a bit more tricky, as we then need to have a method of authenticating to GitHub that we are allowed to download the dependency.** We however cannot use the current flow, as there is no human to enter the code. Github Actions allows some injection of token, I will need to look into this. We would therefore need a mechanism to use this token (an environment variable) inside of the client. This needs to be done in a way that does not impact the current workflow, centered around authenticating inside of the client.

# Building Projects with dependencies in CI

Again, we need to get the dependencies before the build can succeed. I identified three ways of achieving this:

1. The easiest option would be to open a client and load the project in there, essentially creating the same situation as above for the test running. However, this would force us to run the build processes of the lively next installation as well and in general, this solution would need a lot more time than necessary. I would prefer to do this as a last resort only.
2. Maybe this is the moment to fully embrace `flatn`? I think that could be cool, especially since we would end up with proper conflict resolution etc. but it seems overkill to me for now.
3. We could basically copy the resolution mechanism that ensures the existence of all dependencies in the client and put something similar in the bundler as well. Downside is, that the mechanism is quite complex, and we end up with a good amount of code duplication.
4. My preferred solution: We determine the whole dependency-chain at runtime in the client anyways. We also regenerate the pipeline-files at regular points. We could couple changing dependencies of a project to the recreation of the build file, and generate `git clone` steps for each dependency, ensuring their existence without the need to open the client. I think this is the easiest option to implement which also leaves us with the least amount of code in the end.

The credentials problematic remains the same as above, but since we can resolve the problem inside of the pipeline alone in this case, once we find out how to solve the problem inside of GitHub Actions, we are done and do not need to worry about somehow passing this inside of the lively client.

@merryman let's talk about this some time, please. I spent a good amount of time thinking about this, but want to make sure that I did not miss something obvious before proceeding.

**Comment by linusha** [comment]
Part of solving this is also showing useful warnings and hints to end users, when they use private repositories as dependencies!

**Comment by linusha** [comment]
We do four.

**Comment by linusha** [comment]
I investigated how to retrieve the private repositories inside of the Action runs of projects that have them as dependency. I will document my findings here for future reference, although I am positive that I will finish the final implementation quite soon:

I first experimented with a solution that takes the github token that we generate with the lively-authentication app and puts that inside of the repository of the project as a secret. As that token has `repo` scope, we can then clone all other projects that are accessible with that token.
Upon further investigation, this is a bad idea: Collaborators on this project will subsequently be able to use this secret and could in theory gain access to **all** of my repositories by abusing the token stored inside of the shared project repository. The crux here is that github tokens cannot just have a scope limited to a few repositories.

However, we can also use so-called deployment keys, which are basically SSH keys. The setup needs to be the following:
- make the private key available in the project,
- make the public key available to the dependencies.

We can set these secrets up using the API. Thus, when creating a project, we generate a ssh keypair and setup a secret inside of the repo of the project containing the secret key. We can do that using the API. We might need a mechanism that allows us to do that later, but we can always check for the existence of a secret, so that is no problem.
We will also need to store the public key. I actually suggest just storing that somewhere inside of the repo, so that we can always read it later. Should be no problem, as its a public key anyways, right? 
When we add a dependency to a project, check if its private. If thats the case, add the public key as a deploy key. 
We will then generate the approproate github actions, cloning the dependencies using the private key secret and save that in the project.

In case you read this soon, @merryman, any major problems with this that stand out to you? Especially regarding potential security problems?
